README


This archive contains the distribution MooX-Types-MooseLike,
version 0.03:

  Moose Like Types for Moo

This software is copyright (c) 2012 by Mateu X Hunter.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.




Changes
0.03      2012-03-19 10:44:10 America/Denver
          IMPROVEMENTS:
          * Bunches of POD [RT#73519]
          * Simplified Exporter usage [RT#75209] (dolmen)
          * Removed Data::Dumper::Concise dependency [RT#73469] (rsimoes)

          ADDITIONS:
          * Numeric Types
          * Experimental type building API

          API CHANGES:
          * Basic types are now found in MooX::Types::MooseLike::Base
            instead of MooX::Types::MooseLike


0.02      2011-10-21 17:45:17 MST7MDT
          ADDITIONS:
          * Most non-parameterized types of the basic Moose type hierarchy
            are now supported.  Exceptions are ClassName and RoleName
          * Ability to re-use the type tests to build new ones via is_$type()

          IMPROVEMENTS:
          * Refactored generation of assert_Foo() to use a dispatch table
            and captures for quote_sub.

          FIXES:
          * Add AutoPrereq to dist.ini so Makefile.PL gets dependencies generated

0.01      2011-09-15
          Initial Release

lib/MooX/Types/MooseLike.pm
use strictures 1;
package MooX::Types::MooseLike;
{
  $MooX::Types::MooseLike::VERSION = '0.03';
}
use Exporter 5.57 'import';

sub register_types {
  my ($type_definitions, $into) = @_;
  foreach my $type_def (@{$type_definitions}) {
    my $coderefs = make_type($type_def);
    install_type($type_def->{name}, $coderefs, $into);
  }
}

sub install_type {
  my ($type_name, $coderefs, $namespace) = @_;
  my $is_type_name          = 'is_' . $type_name;
  my $type_full_name        = $namespace . '::' . $type_name;
  my $is_type_full_name     = $namespace . '::' . $is_type_name;

  {
    no strict 'refs';    ## no critic
    *{$type_full_name} = $coderefs->{type};
    *{$is_type_full_name} = $coderefs->{is_type};
    push @{"${namespace}::EXPORT_OK"}, $type_name, $is_type_name;
  }
  return;
}

sub make_type {
  my ($type_definition) = @_;
  my $test = $type_definition->{test};
 
  my $full_test = $test; 
  if (my $subtype_of = $type_definition->{subtype_of}) {
    my $from = $type_definition->{from} || die "Must define a 'from' namespace for the parent type: $subtype_of
 when defining type: $type_definition->{name}";
    my $subtype_test = $from . '::is_' . $subtype_of;
    no strict 'refs';    ## no critic
    $full_test = sub {return (&{$subtype_test}(@_) && $test->(@_)); };
  }

  return {
    type    =>  sub {
      sub { die $type_definition->{message}->(@_) if not $full_test->(@_); };
    },
    is_type => sub { $test->($_[0]) },
  };
}

1;
__END__ 

=head1 NAME

MooX::Types::MooseLike - some Moosish types and a typer builder

=head1 SYNOPSIS

    # The Base and Numeric types are stable
    # but the API to build new types is Experimental
    package MyApp::Types;
    use MooX::Types::MooseLike::Base;
    use base qw(Exporter);
    our @EXPORT_OK = ();
    my $defs = [{ 
      name => 'MyType', 
      test => sub { predicate($_[0]) }, 
      message => sub { "$_[0] is not the type we want!" }
    }];
    MooX::Types::MooseLike::register_types($defs, __PACKAGE__);
    # optionally add an 'all' tag so one can:
    # use MyApp::Types qw/:all/; # to import all types
    our %EXPORT_TAGS = ('all' => \@EXPORT_OK);

=head1 DESCRIPTION

See L<MooX::Types::MooseLike::Base> for an example of how to build base types.

See L<MooX::Types::MooseLike::Numeric> for an example of how to build subtypes.




lib/MooX/Types/MooseLike/Base.pm
use strictures 1;

package MooX::Types::MooseLike::Base;
{
  $MooX::Types::MooseLike::Base::VERSION = '0.03';
}
use Scalar::Util qw(blessed);
use List::Util;
use MooX::Types::MooseLike;
use Exporter 5.57 'import';
our @EXPORT_OK = ();

# These types act like those found in Moose::Util::TypeConstraints.
# Generally speaking, the same test is used.
my $type_definitions = [
  {
    name => 'Any',
    test => sub { 1 },
    message =>
      sub { "If you get here you've achieved the impossible, congrats." }
  },
  {
    name => 'Item',
    test => sub { 1 },
    message =>
      sub { "If you get here you've achieved the impossible, congrats" }
  },
  {
    name    => 'Undef',
    test    => sub { !defined($_[0]) },
    message => sub { "$_[0] is not undef" }
  },

# Note the single quotes so $_[0] is not interpreted because undef and strictures => FATAL
  {
    name    => 'Defined',
    test    => sub { defined($_[0]) },
    message => sub { '$_[0] is not defined' }
  },
  {
    name => 'Bool',

    #  	test    => sub { $_[0] == 0 || $_[0] == 1 },
    test =>
      sub { !defined($_[0]) || $_[0] eq "" || "$_[0]" eq '1' || "$_[0]" eq '0' }
    ,
    message => sub { "$_[0] is not a Boolean" },
  },
  {
    name    => 'Value',
    test    => sub { !ref($_[0]) },
    message => sub { "$_[0] is not a value" }
  },
  {
    name    => 'Ref',
    test    => sub { ref($_[0]) },
    message => sub { "$_[0] is not a reference" }
  },
  {
    name => 'Str',
    test => sub { ref(\$_[0]) eq 'SCALAR' },
    message => sub { "$_[0] is not a string" }
  },
  {
    name    => 'Num',
    test    => sub { Scalar::Util::looks_like_number($_[0]) },
    message => sub { "$_[0] is not a Number!" },
  },
  {
    name => 'Int',

#  	test    => sub { Scalar::Util::looks_like_number($_[0]) && ($_[0] == int $_[0]) },
    test => sub { "$_[0]" =~ /^-?[0-9]+$/ },
    message => sub { "$_[0] is not an Integer!" },
  },
  {
    name => 'ScalarRef',
    test => sub { ref($_[0]) eq 'SCALAR' },
    message => sub { "$_[0] is not an ScalarRef!" },
  },
  {
    name => 'ArrayRef',
    test => sub { ref($_[0]) eq 'ARRAY' },
    message => sub { "$_[0] is not an ArrayRef!" },
  },
  {
    name => 'HashRef',
    test => sub { ref($_[0]) eq 'HASH' },
    message => sub { "$_[0] is not a HashRef!" },
  },
  {
    name => 'CodeRef',
    test => sub { ref($_[0]) eq 'CODE' },
    message => sub { "$_[0] is not a CodeRef!" },
  },
  {
    name => 'RegexpRef',
    test => sub { ref($_[0]) eq 'Regexp' },
    message => sub { "$_[0] is not a RegexpRef!" },
  },
  {
    name => 'GlobRef',
    test => sub { ref($_[0]) eq 'GLOB' },
    message => sub { "$_[0] is not a GlobRef!" },
  },
  {
    name => 'FileHandle',
    test => sub {
      Scalar::Util::openhandle($_[0])
        || (blessed($_[0]) && $_[0]->isa("IO::Handle"));
    },
    message => sub { "$_[0] is not a FileHandle!" },
  },
  {
    name => 'Object',
    test => sub { blessed($_[0]) && blessed($_[0]) ne 'Regexp' },
    message => sub { "$_[0] is not an Object!" },
  },
  {
    name => 'AHRef',
    test => sub {
      (ref($_[0]) eq 'ARRAY')
        && ($_[0]->[0])
        && (List::Util::first { ref($_) eq 'HASH' } @{ $_[0] });
    },
    message => sub { "$_[0] is not an ArrayRef[HashRef]!" },
  },
];

MooX::Types::MooseLike::register_types($type_definitions, __PACKAGE__);
our %EXPORT_TAGS = ('all' => \@EXPORT_OK);

1

__END__ 

=head1 NAME

MooX::Types::MooseLike::Base - Moose like types for Moo

=head1 SYNOPSIS

    package MyPackage;
    use Moo;
    use MooX::Types::MooseLike::Base qw(:all);
    
    has "beers_by_day_of_week" => (
        isa => HashRef
    );
    has "current_BAC" => (
        isa => Num
    );
    
    # Also supporting is_$type.  For example, is_Int() can be used as follows
    has 'legal_age' => (
        is => 'ro',
        isa => sub { die "$_[0] is not of legal age" 
        	           unless (is_Int($_[0]) && $_[0] > 17) },
    );

=head1 DESCRIPTION

Moo attributes (like Moose) have an 'isa' property.
This module provides some basic types for this property.
One can import all types with ':all' tag or import
a list of types like:

    use MooX::Types::MooseLike::Base qw/HashRef CodeRef/;

so one could then declare some attributtes like:

	has 'contact' => (
	  is => 'ro',
	  isa => HashRef,
	);
	has 'guest_list' => (
	  is => 'ro',
	  isa => ArrayRef,
	);

These types provide a check that the contact attribute is a hash reference,
and that the guest_list is an array reference.

=head1 TYPES (subroutines)

All available types are listed below.

=head2 Any

Any type (test is always true)

=head2 Item

Synonymous with Any type 

=head2 Undef

A type that is not defined

=head2 Defined

A type that is defined

=head2 Bool

A boolean 1|0 type

=head2 Value

A non-reference type

=head2 Ref

A reference type

=head2 Str

A non-reference type where a reference to it is a SCALAR

=head2 Num

A number type

=head2 Int

An integer type

=head2 ArrayRef

An ArrayRef (ARRAY) type

=head2 HashRef

A HashRef (HASH) type

=head2 CodeRef

A CodeRef (CODE) type

=head2 RegexpRef

A regular expression reference type

=head2 GlobRef

A glob reference type

=head2 FileHandle

A type that is either a builtin perl filehandle or an IO::Handle object

=head2 Object

A type that is an object (think blessed)

=head2 AHRef

An ArrayRef[HashRef] type

=head1 AUTHOR

Mateu Hunter C<hunter@missoula.org>

=head1 THANKS

mst provided the implementation suggestion of using 'can' on a quoted sub 
to define a type (subroutine).

=head1 COPYRIGHT

Copyright 2011, Mateu Hunter

=head1 LICENSE

You may distribute this code under the same terms as Perl itself.

=cut


lib/MooX/Types/MooseLike/Numeric.pm
use strictures 1;

package MooX::Types::MooseLike::Numeric;
{
  $MooX::Types::MooseLike::Numeric::VERSION = '0.03';
}
use MooX::Types::MooseLike::Base;
use Exporter 5.57 'import';
our @EXPORT_OK = ();

my $type_definitions = [
  {
    name       => 'PositiveNum',
    subtype_of => 'Num',
    from       => 'MooX::Types::MooseLike::Base',
    test       => sub { $_[0] > 0 },
    message    => sub { "$_[0] is not a positive number!" },
  },
  {
    name       => 'NonNegativeNum',
    subtype_of => 'Num',
    from       => 'MooX::Types::MooseLike::Base',
    test       => sub { $_[0] >= 0 },
    message    => sub { "$_[0] is not a non-negative number!" },
  },
  {
    name       => 'PositiveInt',
    subtype_of => 'Int',
    from       => 'MooX::Types::MooseLike::Base',
    test       => sub { $_[0] > 0 },
    message    => sub { "$_[0] is not a positive integer!" },
  },
  {
    name       => 'NonNegativeInt',
    subtype_of => 'Int',
    from       => 'MooX::Types::MooseLike::Base',
    test       => sub { $_[0] >= 0 },
    message    => sub { "$_[0] is not a non-negative integer!" },
  },
  {
    name       => 'NegativeNum',
    subtype_of => 'Num',
    from       => 'MooX::Types::MooseLike::Base',
    test       => sub { $_[0] < 0 },
    message    => sub { "$_[0] is not a negative number!" },
  },
  {
    name       => 'NonPositiveNum',
    subtype_of => 'Num',
    from       => 'MooX::Types::MooseLike::Base',
    test       => sub { $_[0] <= 0 },
    message    => sub { "$_[0] is not a non-positive number!" },
  },
  {
    name       => 'NegativeInt',
    subtype_of => 'Int',
    from       => 'MooX::Types::MooseLike::Base',
    test       => sub { $_[0] < 0 },
    message    => sub { "$_[0] is not a negative integer!" },
  },
  {
    name       => 'NonPositiveInt',
    subtype_of => 'Int',
    from       => 'MooX::Types::MooseLike::Base',
    test       => sub { $_[0] <= 0 },
    message    => sub { "$_[0] is not a non-positive integer!" },
  },
  {
    name       => 'SingleDigit',
    subtype_of => 'NonNegativeInt',
    from       => 'MooX::Types::MooseLike::Numeric',
    test       => sub { $_[0] < 10 },
    message    => sub { "$_[0] is not a single digit!" },
  },
];

MooX::Types::MooseLike::register_types($type_definitions, __PACKAGE__);
our %EXPORT_TAGS = ('all' => \@EXPORT_OK);

1

__END__

=head1 NAME

MooX::Types::MooseLike::Numeric - Moo types for numbers

=head1 SYNOPSIS

    package MyPackage;
    use Moo;
    use MooX::Types::MooseLike::Numeric qw(PositiveInt);

    has "daily_breathes" => (
        isa => PositiveInt
    );

=head1 DESCRIPTION

adpated from MooseX::Types::Common::Numeric

=head1 TYPES (subroutines)

Available types are listed below.

=over

=item PositiveNum

=item NonNegativeNum

=item PositiveInt

=item NonNegativeInt

=item NegativeNum

=item NonPositiveNum

=item NegativeInt

=item NonPositiveInt

=item SingleDigit

=back


Makefile.PL

use strict;
use warnings;



use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Moose Like Types for Moo",
  "AUTHOR" => "Mateu X Hunter <hunter\@missoula.org>",
  "BUILD_REQUIRES" => {
    "English" => 0,
    "File::Find" => 0,
    "File::Temp" => 0,
    "IO::Handle" => 0,
    "Moo" => 0,
    "Test::Fatal" => 0,
    "Test::More" => 0,
    "warnings" => 0
  },
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "MooX-Types-MooseLike",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "MooX::Types::MooseLike",
  "PREREQ_PM" => {
    "Exporter" => "5.57",
    "List::Util" => 0,
    "Scalar::Util" => 0,
    "strict" => 0,
    "strictures" => 1
  },
  "VERSION" => "0.03",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



